### PR TITLE
Added hash function to elements of the fundamental group of an extended affine Weyl group

### LINKED ISSUE #36121: Missing __hash__ function for elements of FundamentalGroupOfExtendedAffineWeylGroup
### Steps To Reproduce

1. Build the latest version of sagemath from source (though the issue persists in any version of sagemath I encountered over the past years)
2. Open a sagemath console
3. Create an `ExtendedAffineWeylGroup`, e.g.
```
W = ExtendedAffineWeylGroup('A4')
```
4. Create an element of the corresponding `fundamental_group` or `FW`-group, e.g.
```
fun = W.fundamental_group().an_element()
w = W.FW().an_element()
```
5. Attempt to calculate the `hash`-function of said element, or use it in a data structure that depends on it (e.g. as key in a dictionary)
```
hash(fun)
dictionary = {w: 0}
```

### Expected Behavior

The expression `hash(fun)` should evaluate to some integer which is unique among elements of the same fundamental group. The expression `{w: 0}` should evaluate to a python dictionary with `w` as its key.

The Borovoi fundamental group of an extended affine Weyl group is a finite or infinite cyclic group associated to an extended affine Weyl group. Its elements should be considered as mathematical objects and are hence immutable. Associating additional information to them in a python dictionary is desirable in some applications.

The extended affine Weyl group can be constructed as a semidirect product of the non-extended affine Weyl group (which is canonically a Coxeter group) and the Borovoi fundamental group. Its elements should be considered as mathematical objects and are hence immutable. Associating additional information to them in a python dictionary is desirable in many applications. As of now, one has to convert to a different construction of the extended affine Weyl group to do so, e.g.
```
dictionary = {W.W0P()(w): 0}
```
works. However, this conversion makes the code less readable, and can be very time-consuming for large groups and large elements.

### Actual Behavior

```
TypeError: unhashable type: 'FundamentalGroupOfExtendedAffineWeylGroup_Class_with_category.element_class'
```

### Additional Information

Adding the obvious two lines of code to `sage/combinat/root_system/fundamental_group.py` fixes the issue. I will try to submit a corresponding pull request.

### Environment

```markdown
- **OS**: Linux 6.4.0-2-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.4.4-3 (2023-08-08) x86_64 GNU/Linux
- **Sage Version**: SageMath version 10.1, Release Date: 2023-08-20. Using Python 3.11.4.
```

### Checklist

- [X] I have searched the existing issues for a bug report that matches the one I want to file, without success.
- [X] I have read the documentation and troubleshoot guide

### PR DESCRIPTION
**Pull request by 65537**
This change allows users to compute the python `hash` function for elements of the (Borovoi) fundamental group of an extended affine Weyl group, and subsequently for elements of the extended Weyl group itself if using the `FW` realization. This, in turn, allows to use these mathematical objects as keys in dictionaries.

Fixes #36121 . It adds a `__hash__` function to the class `FundamentalGroupElement` associated with `FundamentalGroupOfExtendedAffineWeylGroup`.

SAGE already associates to each element of the fundamental group an integer, accessible as its `value()` or `_value`.  This integer is unique among other elements of the same fundamental group and thus suitable as a hash for hashmap applications (e.g. Python dictionaries). We return the `hash` of that integer, which is the integer itself for current python implementations.

```
W = ExtendedAffineWeylGroup('A4')
fun = W.fundamental_group().an_element()
w = W.FW().an_element()

hash(fun), {w: 0}
```
yields the following result:
```
(4, {pi[4] * S0*S1*S2*S3*S4: 0})
```

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/combinat/root_system/fundamental_group.py [hunk 1]
```diff
@@ -323,6 +323,8 @@ def act_on_affine_lattice(self, wt):
             Doesn't work on ambient spaces.
         """
         return wt.map_support(self.parent().action(self.value()))
+    def __hash__(self):
+        return hash(self.value())
```
Comment (tscrim) [inline src/sage/combinat/root_system/fundamental_group.py]:
```suggestion
        return wt.map_support(self.parent().action(self.value()))

    def __hash__(self):
        r"""
        Return the hash of ``self``.

        TESTS:

        Check that :issue:`36121` is fixed::

            sage: W = ExtendedAffineWeylGroup('A4')
            sage: fun = W.fundamental_group().an_element()
            sage: w = W.FW().an_element()
            sage: hash(w) == hash(w.value())
            True
        """
        return hash(self.value())
```

**Review by tscrim — CHANGES_REQUESTED** [review]
Thank you for noticing and fixing this.

For next time, please be sure to add a doctest showing the issue is fixed (also, all methods need to have at least one doctest).

**Comment by 65537** [comment]
Dear Travis,

thank you for reviewing and improving my pull request. I was following the (mathematically very similar) class WeylGroupElement whose hash function has neither a docstring nor a test, but I will make sure to always include them in future pull requests.

Best, Felix

**Review by tscrim — APPROVED** [review]
Indeed, that is generally a good idea. Unfortunately in this case that is some fairly old code that was included before we developed our modern practices. Thank you for adding it. LGTM.

**Comment by vbraun** [comment]
```
sage -t --long --warn-long 39.2 --random-seed=123 src/sage/combinat/root_system/fundamental_group.py
**********************************************************************
File "src/sage/combinat/root_system/fundamental_group.py", line 339, in sage.combinat.root_system.fundamental_group.FundamentalGroupElement.__hash__
Failed example:
    hash(w) == hash(w.value())
Exception raised:
    Traceback (most recent call last):
      File "/home/release/Sage/src/sage/doctest/forker.py", line 709, in _run
        self.compile_and_execute(example, compiler, test.globs)
      File "/home/release/Sage/src/sage/doctest/forker.py", line 1144, in compile_and_execute
        exec(compiled, globs)
      File "<doctest sage.combinat.root_system.fundamental_group.FundamentalGroupElement.__hash__[3]>", line 1, in <module>
        hash(w) == hash(w.value())
                        ^^^^^^^^^
    TypeError: 'tuple' object is not callable
**********************************************************************
```

**Comment by tscrim** [comment]
@65537 Sorry, I didn’t check the doctest carefully. Please change `value()` to `value`.

**Review by 65537 — COMMENTED** [review]
Testing for `value()` is correct, as long as you are testing on `FundamentalGroupElement`s, which was the main culprit here

**Comment by tscrim** [comment]
Ah, indeed.

**Comment by tscrim** [comment]
@vbraun Has this been merged? If so, should we close this?

**Pull request merged**